lib/partition_healing/healer.js

```diff
@@ -0,0 +1,193 @@
+ * Healer implements an algorithm to heal a partitioned ringpop cluster.
+ *
+ * @param ringpop
+ *
+ * @constructor
+ */
+function Healer(ringpop) {
+    this.ringpop = ringpop;
+    this.logger = this.ringpop.loggerFactory.getLogger('healer');
+}
+
+
+/**
+ * Attempt a heal between the current and target node. Calling this function does not
+ * result in a heal ff there are nodes that need to be reincarnated to take precedence
+ * over the faulty declarations. A cluster may need some time and multiple heal attempts
+ * before it is successfully healed.
+ *
+ * @param {string} target The address of the target node of the heal-attempt.
+ * @param {Healer~attemptHealCallback} callback called when the heal attempt is done.
+ * @protected
+ */
+Healer.prototype.attemptHeal = function attemptHeal(target, callback) {
+    var self = this;
+
+    var remoteMembership = null;
+    
+    async.waterfall([
+        sendJoinRequest,
+        generateChanges,
+        processChanges
+    ], function done(err) {
+        if (err) {
+            callback(err);
+            return;
+        }
+
+        var pingableHosts = _.chain(remoteMembership).filter(function isPingable(change) {
+            return Member.isStatusPingable(change.status);
+        }).pluck('address').value();
+
+        callback(null, pingableHosts);
+    });
+
+    function sendJoinRequest(next) {
+        self.ringpop.client.protocolJoin({
+            host: target,
+            retryLimit: self.ringpop.config.get('tchannelRetryLimit'),
+            timeout: 1000
+        }, {
+            app: self.ringpop.app,
+            source: self.ringpop.whoami(),
+            incarnationNumber: self.ringpop.membership.localMember.incarnationNumber
+        }, next);
+    }
+
+    function generateChanges(joinResponse, next) {
+        remoteMembership = joinResponse.membership;
+
+        // Index local membership by address for faster lookups.
+        var localMembershipMap = _.indexBy(self.ringpop.dissemination.fullSync(), 'address');
+
+        var localChanges = [];
+        var remoteChanges = [];
+
+        for (var i = 0; i < remoteMembership.length; i++) {
+            var remoteMember = remoteMembership[i];
+            var localMember = localMembershipMap[remoteMember.address];
+
+            if (!localMember) {
+                continue;
+            }
+
+            // if the local member would become un-pingable after applying the remote change
+            if (nodeWouldBecomeUnPingable(localMember, remoteMember)) {
+                // mark it locally as suspect
+                localChanges.push(_.extend({}, remoteMember, {status: Member.Status.suspect}));
+            }
+
+            // if the remote member would become un-pingable after applying the local change
+            if (nodeWouldBecomeUnPingable(remoteMember, localMember)) {
+                // mark it remotely as suspect
+                remoteChanges.push(_.extend({}, localMember, {status: Member.Status.suspect}));
```

Comment:
since the local changes are created from the `fullsync` call they contain `source` information about this node if I remember correctly. This in combination with bidirectional fullsync's caused an issue in the ringpop-node implementation.

I think we solved it by scrubbing (or never putting it on there) the source information from the change we sent remote.

If we don't do so the remote might initiate a fullsync back to this node when it receives the 'reincarnation' ping during the partition healing causing the 'bridge' between the two partitions to be created too soon.